src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs

```diff
@@ -419,7 +419,7 @@ private async ValueTask<FlushResult> WriteDataAsync(int streamId, StreamOutputFl
                 // It should be very rare for a continuation to run without any availability.
                 if (availabilityAwaitable != null)
                 {
-                    await availabilityAwaitable;
+                    await new ValueTask<object>(availabilityAwaitable, availabilityAwaitable.Version);
```

Comment:
I think it would be better to create the ValueTask when we still have the `_writeLock`. If there's ever a pooling bug in the future, that would make the token more likely to have changed by the time ValueTask.OnCompleted and GetResult() are called which should help us catch the issue earlier.

Reply:
Done. AdvanceUpToAndWait now returns a ValueTask. Feels cleaner than returning an awaitable.